feat(ast-spec): add fixture test framework and some initial fixtures

bradzacher opened the pull request:
Add fixture tests for each and every AST node.
To make things more readable I:
- separated the AST from the tokens
  - this stops the individual snapshot from being so long it's impossible to understand
  - if there are changes to tokens - then they should be more obvious now. Before you had to expand the snapshot in github to understand that the token was part of the tokens array.
- always write error snapshot, and always write AST/token snapshots
  - This is just for consistency and to make it easier whilst you're developing.
  - It will prevent us from accidentally leaving behind error snapshots if they weren't supposed to be there.
- added a custom snapshot serializer
  - Having learned how useful they are with scope manager - I created a new one to improve the look of the snapshots.
  - Instead of sorting alphabetically, I place `type` at the top, and `range`/`loc` at the end.
  - Instead of outputting `Object` ahead of every node, instead it outputs the node `type`.
  - I adjusted the output `range`/`loc` so they take up fewer lines and are more compact.

I prefixed the snapshot names with numbers just so we can control the sorting of them. No other reason.

# Work In Progress

I wanted to get some early feedback on things like the folder structure and the snapshot representation before I go too far down this rabbit hole.

TODO:
- [ ] Add fixtures for every single node
- [x] Add babel parse snapshots
- [x] Add babel alignment snapshots

bradzacher commented on packages/ast-spec/src/declaration/ClassDeclaration/fixtures/implementsMany/snapshots/7-AST-Alignment-AST.shot:
```diff
@@ -0,0 +1,115 @@
+-         TSClassImplements {
+-           type: 'TSClassImplements',
++         TSExpressionWithTypeArguments {
++           type: 'TSExpressionWithTypeArguments',
+            expression: Identifier {
+              type: 'Identifier',
+              name: 'RegExp',
+
+              range: [39, 45],
+              loc: {
+                start: { column: 39, line: 1 },
+                end: { column: 45, line: 1 },
+              },
+            },
+
+            range: [39, 45],
+            loc: {
+              start: { column: 39, line: 1 },
+              end: { column: 45, line: 1 },
+            },
+          },
+        ],
+        superClass: null,
+
+        range: [0, 48],
+        loc: {
+          start: { column: 0, line: 1 },
+          end: { column: 48, line: 1 },
+        },
+      },
+    ],
+    sourceType: 'script',
+
+    range: [0, 49],
+    loc: {
+      start: { column: 0, line: 1 },
+      end: { column: 0, line: 2 },
+    },
+  }"
+`;
```
example of an AST misalignment.

[**Rendered in GH**](https://github.com/typescript-eslint/typescript-eslint/blob/63f5b1ab8a27cb0daf12a325988ffe674c3267be/packages/ast-spec/src/declaration/ClassDeclaration/fixtures/implementsMany/snapshots/7-AST-Alignment-AST.shot)

bradzacher commented on packages/ast-spec/src/declaration/ClassDeclaration/fixtures/implementsMany/snapshots/8-AST-Alignment-Tokens.shot:
```diff
@@ -0,0 +1,112 @@
+      value: ',',
+
+      range: [37, 38],
+      loc: {
+        start: { column: 37, line: 1 },
+        end: { column: 38, line: 1 },
+      },
+    },
+    Identifier {
+      type: 'Identifier',
+      value: 'RegExp',
+
+      range: [39, 45],
+      loc: {
+        start: { column: 39, line: 1 },
+        end: { column: 45, line: 1 },
+      },
+    },
+    Punctuator {
+      type: 'Punctuator',
+      value: '{',
+
+      range: [46, 47],
+      loc: {
+        start: { column: 46, line: 1 },
+        end: { column: 47, line: 1 },
+      },
+    },
+    Punctuator {
+      type: 'Punctuator',
+      value: '}',
+
+      range: [47, 48],
+      loc: {
+        start: { column: 47, line: 1 },
+        end: { column: 48, line: 1 },
+      },
+    },
+  ]"
+`;
```
example of an token misalignment.

[**Rendered in GH**](https://github.com/typescript-eslint/typescript-eslint/blob/63f5b1ab8a27cb0daf12a325988ffe674c3267be/packages/ast-spec/src/declaration/ClassDeclaration/fixtures/implementsMany/snapshots/8-AST-Alignment-Tokens.shot)

bradzacher commented on packages/ast-spec/tests/fixtures-with-differences-ast.shot:
```diff
@@ -0,0 +1,10 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`AST Fixtures List fixtures with AST differences 1`] = `
+Set {
+  "src/declaration/ClassDeclaration/fixtures/implementsMany/fixture.ts",
+  "src/declaration/ClassDeclaration/fixtures/implementsOne/fixture.ts",
+  "src/declaration/ClassDeclaration/fixtures/typeParameters/fixture.ts",
+  "src/declaration/ClassDeclaration/fixtures/typeParametersExtendsTypeParameters/fixture.ts",
+}
+`;
```
snapshot which provides a centralised list of fixtures with AST differences

bradzacher commented on packages/ast-spec/tests/fixtures-with-differences-tokens.shot:
```diff
@@ -0,0 +1,8 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`AST Fixtures List fixtures with Token differences 1`] = `
+Set {
+  "src/declaration/ClassDeclaration/fixtures/implementsMany/fixture.ts",
+  "src/declaration/ClassDeclaration/fixtures/implementsOne/fixture.ts",
+}
+`;
```
snapshot which provides a centralised list of fixtures with Token differences

bradzacher commented on packages/ast-spec/tests/fixtures-with-differences-errors.shot:
```diff
@@ -0,0 +1,16 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`AST Fixtures List fixtures with Error differences 1`] = `
+Object {
+  "Babel errored but TSESTree didn't": Set {
+    "declaration/ClassDeclaration/fixtures/_error_/implements-non-identifier/fixture.ts",
+    "declaration/ExportAllDeclaration/fixtures/_error_/non-string-source/fixture.ts",
+    "declaration/ExportAllDeclaration/fixtures/_error_/type-kind/fixture.ts",
+    "declaration/ExportNamedDeclaration/fixtures/_error_/anonymous-class/fixture.ts",
+  },
+  "TSESTree errored but Babel didn't": Set {
+    "declaration/ExportAllDeclaration/fixtures/_error_/named-non-identifier/fixture.ts",
+    "declaration/ExportNamedDeclaration/fixtures/_error_/aliased-literal/fixture.ts",
+  },
+}
+`;
```
snapshot which provides a centralised list of fixtures with error differences (not error messages, just existence of errors)

JoshuaKGoldberg commented:
Sure! I'll try to find time Soon™️.

JoshuaKGoldberg commented on packages/ast-spec/src/declaration/ExportNamedDeclaration/spec.ts:
```diff
@@ -7,6 +7,7 @@ import type { ExportKind } from '../ExportAndImportKind';
 
 export interface ExportNamedDeclaration extends BaseNode {
   type: AST_NODE_TYPES.ExportNamedDeclaration;
+  // TODO(error handling) - ClassExpression is not valid here
```
Are there tracking issue for these todos?

bradzacher replied:
#1852 is the issue for all  the "we should add AST errors" todos
I can update these comments with the issue number for reference

JoshuaKGoldberg commented on packages/ast-spec/tests/util/setupJest.ts:
```diff
@@ -0,0 +1,8 @@
+import { serializers } from './serializers';
+
+for (const serializer of serializers) {
+  // the jest types are wrong here
```
oh?

bradzacher replied:
I actually think this is fixed now

JamesHenry replied:
This was marked as resolved a some time ago, but the comment still remains in the code so wanted to open it up again to a final resolution now

JoshuaKGoldberg commented on packages/ast-spec/tests/util/serializers/index.ts:
```diff
@@ -0,0 +1,6 @@
+import * as Node from './Node';
+
+export const serializers = [
+  //
```
```suggestion
```

JoshuaKGoldberg reviewed (commented):
Yeah I like this a _lot_. Always a +1 from me for splitting up big files!

Just a few curiosities but recognized this is a draft.

bradzacher commented:
The core of this is done - all that's left is creating fixtures for each and every case.
Instead of holding it in draft until I find the time to do this - I'm going to put this up for review so we can merge it.

That way we can make all new AST changes add fixtures here, and migrate the old fixtures piece-by-piece - we can distribute the work so I'm not the blocker.

JoshuaKGoldberg commented on packages/ast-spec/tests/fixtures.test.ts:
```diff
@@ -0,0 +1,249 @@
+  },
+);
+
+function isError(e: unknown): e is { code: unknown } {
```
Nit: I would have thought `isError` is checking for `e is Error`. Maybe...

```suggestion
function hasErrorCode(e: unknown): e is { code: unknown } {
```

JoshuaKGoldberg commented on packages/ast-spec/tests/util/parsers/babel.ts:
```diff
@@ -0,0 +1,50 @@
+    plugins.push('jsx');
+  }
+
+  let ast: unknown | 'ERROR' = 'ERROR';
```
Suggestion: maybe use an enum for `'ERROR'` and `'NO ERROR'`? These constants are duplicated quite a few times.

JoshuaKGoldberg commented on packages/ast-spec/tests/util/parsers/babel.ts:
```diff
@@ -0,0 +1,50 @@
+    plugins.push('jsx');
+  }
+
+  let ast: unknown | 'ERROR' = 'ERROR';
+  let tokens: unknown | 'ERROR' = 'ERROR';
+  let error: unknown | 'NO ERROR' = 'NO ERROR';
```
https://typescript-eslint.io/rules/no-redundant-type-constituents for `unknown`:

```suggestion
  let ast: unknown = 'ERROR';
  let tokens: unknown = 'ERROR';
  let error: unknown = 'NO ERROR';
```

JoshuaKGoldberg commented on packages/ast-spec/tests/util/serializers/string.ts:
```diff
@@ -0,0 +1,38 @@
+    characters.push("'");
+
+    return characters.join('');
+  },
```
This functionally behaves about the same as `str::replaceAll("'", '\\')`, no?

bradzacher replied:
I think you could technically write this as
```ts
str.replaceAll("\\", "\\\\").replaceAll("'", "\\'");
```
Though that would be slower because you'd need to do it in two passes over the string.

I haven't perf tested it, but given that there are going to be hundreds of these strings - each with a few hundred characters, I think that double handling will add up.

JoshuaKGoldberg commented on packages/ast-spec/tests/util/typescript-estree-import.ts:
```diff
@@ -0,0 +1,17 @@
+ * - Removing @typescript-eslint/typescript-estree as an explicit devDependency in the package.json
+ * - Add an `.nxignore` file at the root of the monorepo which ignores this file (which we then in turn
+ * ensure is the only place we directly import from the @typescript-eslint/typescript-estree package).
+ */
```
This feels like something we (@JamesHenry 😄) should feature request into Nx: the ability to have technical circular dependencies only in tests...

JamesHenry replied:
I can't remember the exact history of this but I'm pretty sure I provided this breakdown and created this solution to the problem.

As it turns out on this PR right now the `@typescript-eslint/typescript-estree` package is not in the package.json of ast-spec, so we could simplify things to:

<img width="1512" alt="image" src="https://user-images.githubusercontent.com/900523/164988392-53c7d8fa-def1-4b88-9a60-194ef93904a0.png">

The specifying of the package in the package.json is the thing that causes the rigid circular dependency to exist - there would be no way to know that the package was only used in certain situations when that is the way the relationship is defined. As part of the Nx updates I'm doing we'll be moving entirely to source to source relationships between packages/graph nodes (as is far more common) and we won't need to worry about this in that situation

JamesHenry replied:
(I didn't include it in the screenshot but feel free to move some of the commentary around why we _don't_ want to include the package in the package.json to make that lint rule happy and that's why we are disabling it inline)

bradzacher replied:
Was there an update to nx to allow this? I swear there was a reason we went with using an `.nxignore` file.

Or did we used to rely on implicit dependency graph and now we have an explicit one configured it doesn't matter?

JoshuaKGoldberg commented on packages/ast-spec/typings/babel-eslint-parser.d.ts:
```diff
@@ -0,0 +1,24 @@
+  }
+
+  export function parse(code: string, options: Options): BabelAST;
+  // export function parseForESLint(code: string, options: Options): unknown;
```
```suggestion
```

JoshuaKGoldberg commented on yarn.lock:
```diff
@@ -137,18 +137,46 @@
+  dependencies:
+    "@babel/highlight" "^7.14.5"
+
+"@babel/code-frame@^7.10.4", "@babel/code-frame@^7.16.0", "@babel/code-frame@^7.16.7", "@babel/code-frame@^7.8.3":
```
You might want to `npx yarn-deduplicate`; these are starting to get more splintered.

bradzacher replied:
`yarn-deduplicate` will only consolidate the versions of a package.
In this case this line is saying there are a bunch of packages which all need code frame and have different semver specifiers, but they're all resolving to exactly the same version!

JoshuaKGoldberg commented on packages/types/tools/copy-ast-spec.ts:
```diff
@@ -1,11 +1,39 @@
+
+    //   // console.log(stdout);
+    //   resolve();
+    // },
```
Delete?

JamesHenry replied:
One minor thing if you're going to update this util  anyway to delete the code - one thing that can be nasty to troubleshoot if it ever comes up is the stingy `maxBuffer` size that node gives you by default.

`execa`, for example, ups this to 100MB https://github.com/sindresorhus/execa/blob/fa9cc053837a3c78f61ee1f330aa4048bb8d278f/index.js#L16 so we might want to do something similar

bradzacher replied:
In the case of `child_process.exec` - it is built to execute the command, buffer the entire output, then return it as a string.
This is why a buffer is important for it - because it buffers the output and then returns it.

For this I've used `spawn` with `stdio: 'inherit'`.
This means that the log lines will just be streamed directly to the main process's stdout and thus there's no buffer collecting the output for an extended period.
In fact, `spawn` has no option to control the buffer size!

JoshuaKGoldberg reviewed (approved):
🔥. Just a few small code nitpicks from me on re-review, nothing close to blocking. Awesome!!

I'd want to hear from @JamesHenry on the Nx shenanigans, if James has time.

JamesHenry commented:
@JoshuaKGoldberg @bradzacher I’ve set aside some time tomorrow 👍

JamesHenry commented on .github/workflows/ci.yml:
```diff
@@ -94,6 +94,12 @@ jobs:
+        run: yarn test
+        working-directory: packages/ast-spec
+        env:
+          CI: true
```
Yeah for sure, once I have finished my Nx refactoring we can opt into the Distributed Task Execution and let Nx Cloud decide how to parallelize and distribute across all available agents, then collect all artifacts

JamesHenry commented on packages/ast-spec/src/declaration/ClassDeclaration/fixtures/extends-literal/README.md:
```diff
@@ -0,0 +1 @@
+This is dumb code - but it's 100% syntactically valid and adheres to the ESTree spec.
```
![download](https://user-images.githubusercontent.com/900523/164984135-051329df-f666-40f5-8ce8-e68e83218e30.jpeg)

JamesHenry commented on packages/ast-spec/tests/util/serializers/Node.ts:
```diff
@@ -0,0 +1,83 @@
+  return typeof type === 'string';
+}
+
+const serializer: NewPlugin = {
```
Love this 👏

JamesHenry commented on packages/ast-spec/tests/util/snapshot-diff.ts:
```diff
@@ -0,0 +1,67 @@
+      // by default jest will quote the string with double quotes
+      // this means the diff string will have double quotes escaped and look ugly
+      // this is a single-quote string serializer which won't clash with the outer double quotes
+      // so we get a nicer looking diff because of it!
```
👍

JamesHenry commented on packages/ast-spec/tests/util/snapshot-diff.ts:
```diff
@@ -0,0 +1,67 @@
+}
+
+// https://github.com/facebook/jest/blob/a293b75310cfc209713df1d34d243eb258995316/packages/jest-diff/src/constants.ts#L8
+const NO_DIFF_MESSAGE = 'Compared values have no visual difference.';
```
Was it not possible to deep import this?

bradzacher replied:
It was until their latest major where they added package.json exports :(

JamesHenry replied:
Ah right fair enough

bradzacher commented:
addressed comments.

side note: I was thinking that this could be a platform for us to work on #1852.
we could use the fixtures to generate an "expected" AST shape and compare it to our types to determine where the types don't match the values.

we also could be running alignment tests against `espree` to help us identify and track where we diverge from the core ESTree spec

JamesHenry commented:
Great work, @bradzacher!

armano2 commented on packages/types/tools/copy-ast-spec.ts:
```diff
@@ -1,11 +1,28 @@
 const writeFile = promisify(fs.writeFile);
-const execAsync = promisify(chlidProcess.exec);
+
+// the promisify util will eat the stderr logs
+async function execAsync(
+  command: string,
+  args: ReadonlyArray<string>,
+  options: childProcess.SpawnOptions,
+): Promise<void> {
+  return new Promise((resolve, reject) => {
+    const child = childProcess.spawn(command, args, {
+      ...options,
+      stdio: 'inherit',
+    });
+
+    child.on('error', e => reject(e));
+    child.on('exit', () => resolve());
+    child.on('close', () => resolve());
+  });
+}
```
@bradzacher this seem to not be working on windows machines

```
Error: spawn yarn ENOENT
    at Process.ChildProcess._handle.onexit (node:internal/child_process:283:19)
    at onErrorNT (node:internal/child_process:478:16)
    at processTicksAndRejections (node:internal/process/task_queues:83:21) {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'spawn yarn',
  path: 'yarn',
  spawnargs: [ 'build' ]
}
```

and rolling this code back to `const execAsync = promisify(childProcess.exec);` seem to be fixing it, not sure what is actually causing this issue

bradzacher replied:
If you can figure out how to make it work on Windows and display the error logs - happy to accept a PR!

I'd assume the path isn't properly set so the binary can't be found? That's generally what EONET means I think.
